### ISSUE TITLE
Find References not working

### ISSUE THREAD

**Issue by gombosg** [scalameta/metals-vscode#765]
**Describe the bug**

**To Reproduce** Steps to reproduce the behavior:

1. Import project
2. Try a "find references" feature
3. No references found anywhere.

**Expected behavior**

Find references is working!

**Screenshots**

Run doctor says it's all right:
![image](https://user-images.githubusercontent.com/23086060/142980887-a413070f-a9b8-40b9-818e-0c3764d93038.png)
![image](https://user-images.githubusercontent.com/23086060/142981506-413cf14a-db43-481e-ba67-6d668048d696.png)

**Installation:**

- Operating system: Fedora 35
- VSCode version: 1.62.3 (bug is present with insiders, too)
- VSCode extension version: 1.10.15
- Metals version: 0.10.9

**Additional context**

[metals.log](https://github.com/scalameta/metals-vscode/files/7586189/metals.log)

...I can also send SemanticDB files if that helps.

Happens in VSCode main and insiders, tried re-importing, cleaning all data etc. it just won't work. I'd so much use Metals for work as it's blazing fast for code navigation, but this is kind of a deal breaker.
Other smaller projects are fine but I don't know what's the key to reproducing this issue in this project.

**Search terms**

find all references

**Comment by tgodzik** [comment]
Thanks for reporting! 

Is it broken for all methods/values etc. in the file? Does it find references of `widget` maybe?

Is the file correctly identified to be a part of a module (it should show type errors from Bloop in that case for example when writing `val a: Int = ""`)

I don't see any issues in the logs regarding references, which should show up if Metals is unable to find semanticdb files for references.

**Comment by gombosg** [comment]
Yes, after investigating, "find all references" works for local `val` variables. But looks like it breaks on function boundaries.
Yes, it throws type errors for `val a: Int = ""`

Is there anywhere else to look for clues?

**Comment by tgodzik** [comment]
Is there anything specific about the render function? Is it maybe an implementation of something from super type? In that case references should only show up on the super type method.

It would be super useful to see if there is any pattern on why some symbols might not be working. If you could send me the semanticdb file it would probably be also useful (for the file with references to `render`). You can read it via `Metals: Show Semanticdb detailed` command.

**Comment by gombosg** [comment]
Definitely! Because find references doesn't work _anywhere_ in the repo (except for local variables), this should be a generic issue.
The other way, it works. So for example, there is a usage of this `render` function in a file called `SlackWidgetRenderService` (also attached). If I ctrl-click on it, jump to definition works fine.
Any more logging I could turn on?

[SlackWidgetRenderService.scala.semanticdb-detailed.txt](https://github.com/scalameta/metals-vscode/files/7594371/SlackWidgetRenderService.scala.semanticdb-detailed.txt)
[CrossLinkWidgetRenderService.scala.semanticdb-detailed.txt](https://github.com/scalameta/metals-vscode/files/7594376/CrossLinkWidgetRenderService.scala.semanticdb-detailed.txt)

**Comment by tgodzik** [comment]
Everything looks good in those two file :thinking: and go to definition also uses semanticdb to get the destination. Nothing new popped up in `metals.log` ?

The only other thing I can think of is that maybe we are wrongly thinking that the references are in an unrelated build target. Is it located in the same module (they look like they are)?

Are the `root` and `root-test` modules correct? Is your app one module only with tests?

I am afraid there might not be much I can do without looking at the project :/ 

Only other option is try to add more debugging options to Metals for the next release.

**Comment by gombosg** [comment]
Hi, yes, it's only these two `root` and `root-test` modules. Let me know if you add more debugging so that I can enable them!

**Comment by yaoqiang-se** [comment]
I have the same problems on windows with metals v0.10.9, but it works fine before when upgrade to v0.10.9.

**Comment by tgodzik** [comment]
> I have the same problems on windows with metals v0.10.9, but it works fine before when upgrade to v0.10.9.

Do you have maybe a reproducible issue in a project? I am having issues figuring out what is happening.

**Comment by yaoqiang-se** [comment]
> > I have the same problems on windows with metals v0.10.9, but it works fine before when upgrade to v0.10.9.
> 
> Do you have maybe a reproducible issue in a project? I am having issues figuring out what is happening.

this is metals doctor:
![image](https://user-images.githubusercontent.com/71065764/146299050-354e9a7a-a934-4e8a-90f6-d47c81ce8eb4.png)

this is issue screenshot:
![image](https://user-images.githubusercontent.com/71065764/146299391-6965daac-34bd-4eeb-bb16-18cb6f4ec1b0.png)

details logs:
Java home: C:\Users\xxx\scoop\apps\openjdk11\current
Metals version: 0.10.9
2021.12.16 10:44:05 INFO  logging to file C:\Users\xxx\Workspace\Code\Lisa\.metals\metals.log
2021.12.16 10:44:05 INFO  Started: Metals version 0.10.9 in workspace 'C:\Users\xxx\Workspace\Code\Lisa' for client Visual Studio Code 1.63.1.
2021.12.16 10:44:07 INFO  time: initialize in 1.81s
Initializing Scala Debugger
2021.12.16 10:44:12 INFO  Flyway Community Edition 8.0.2 by Redgate
2021.12.16 10:44:12 INFO  Database: jdbc:h2:file:C:\Users\xxx\Workspace\Code\Lisa\.metals\metals (H2 1.4)
2021.12.16 10:44:12 INFO  Successfully validated 3 migrations (execution time 00:00.021s)
2021.12.16 10:44:12 INFO  Current version of schema "PUBLIC": 3
2021.12.16 10:44:12 INFO  Schema "PUBLIC" is up to date. No migration necessary.
2021.12.16 10:44:12 INFO  Attempting to connect to the build server...
2021.12.16 10:44:12 INFO  skipping build import with status 'Cancelled'
2021.12.16 10:44:13 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at c:\Users\xxx\Workspace\Code\Lisa\null\scalameta\metals\cache\bsp.trace.json
2021.12.16 10:44:13 INFO  Attempting to connect to the build server...
2021.12.16 10:44:13 INFO  Attempting to connect to the build server...
2021.12.16 10:44:14 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at c:\Users\xxx\Workspace\Code\Lisa\null\scalameta\metals\cache\bsp.trace.json
2021.12.16 10:44:14 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at c:\Users\xxx\Workspace\Code\Lisa\null\scalameta\metals\cache\bsp.trace.json
2021.12.16 10:44:14 INFO  time: Connected to build server in 2.4s
2021.12.16 10:44:14 INFO  Connected to Build server: Bloop v1.4.11
2021.12.16 10:44:16 INFO  no build target: using presentation compiler with only scala-library: 3.1.0
2021.12.16 10:44:16 INFO  time: code lens generation in 4.67s
2021.12.16 10:44:17 INFO  time: Imported build in 0.11s
2021.12.16 10:44:27 INFO  time: indexed workspace in 10s
2021.12.16 10:44:29 INFO  no build target: using presentation compiler with only scala-library: 2.13.6
2021.12.16 10:47:38 INFO  running 'C:\Users\xxx\scoop\apps\openjdk11\current\bin\java -Djline.terminal=jline.UnsupportedTerminal -Dsbt.log.noformat=true -Dfile.encoding=UTF-8 -jar C:\Users\xxx\AppData\Local\Temp\metals7722134289585645032\sbt-launch.jar -Dbloop.export-jar-classifiers=sources bloopInstall'
2021.12.16 10:47:40 INFO  [info] welcome to sbt 1.5.6 (Oracle Corporation Java 11.0.2)
2021.12.16 10:47:41 INFO  [info] loading global plugins from C:\Users\xxx\.sbt\1.0\plugins
2021.12.16 10:47:42 INFO  [info] loading settings for project lisa-build-build-build from metals.sbt ...
2021.12.16 10:47:42 INFO  [info] loading project definition from C:\Users\xxx\Workspace\Code\Lisa\project\project\project
2021.12.16 10:47:42 INFO  [info] loading settings for project lisa-build-build from metals.sbt ...
2021.12.16 10:47:42 INFO  [info] loading project definition from C:\Users\xxx\Workspace\Code\Lisa\project\project
2021.12.16 10:47:45 INFO  [success] Generated .bloop\lisa-build-build.json
2021.12.16 10:47:45 INFO  [success] Total time: 2 s, completed Dec 16, 2021, 10:47:45 AM
2021.12.16 10:47:45 INFO  [info] loading settings for project lisa-build from metals.sbt,plugins.sbt ...
2021.12.16 10:47:45 INFO  [info] loading project definition from C:\Users\xxx\Workspace\Code\Lisa\project
2021.12.16 10:47:52 INFO  [success] Generated .bloop\lisa-build.json
2021.12.16 10:47:52 INFO  [success] Total time: 7 s, completed Dec 16, 2021, 10:47:52 AM
2021.12.16 10:47:53 INFO  [info] loading settings for project root from build.sbt ...
2021.12.16 10:47:53 INFO  [info] set current project to lisa (in build file:/C:/Users/xxx/Workspace/Code/Lisa/)
2021.12.16 10:48:06 INFO  [success] Generated .bloop\root.json
2021.12.16 10:48:06 INFO  [success] Generated .bloop\root-test.json
2021.12.16 10:48:06 INFO  [success] Total time: 13 s, completed Dec 16, 2021, 10:48:07 AM
2021.12.16 10:48:07 INFO  time: ran 'sbt bloopInstall' in 28s
2021.12.16 10:48:07 INFO  Disconnecting from Bloop session...
2021.12.16 10:48:07 INFO  Shut down connection with build server.
2021.12.16 10:48:07 INFO  Shut down connection with build server.
2021.12.16 10:48:07 INFO  Shut down connection with build server.
2021.12.16 10:48:07 INFO  Attempting to connect to the build server...
2021.12.16 10:48:07 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at c:\Users\xxx\Workspace\Code\Lisa\null\scalameta\metals\cache\bsp.trace.json
2021.12.16 10:48:07 INFO  Attempting to connect to the build server...
2021.12.16 10:48:07 INFO  Attempting to connect to the build server...
2021.12.16 10:48:08 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at c:\Users\xxx\Workspace\Code\Lisa\null\scalameta\metals\cache\bsp.trace.json
2021.12.16 10:48:08 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at c:\Users\xxx\Workspace\Code\Lisa\null\scalameta\metals\cache\bsp.trace.json
2021.12.16 10:48:08 INFO  time: Connected to build server in 0.82s
2021.12.16 10:48:08 INFO  Connected to Build server: Bloop v1.4.11
2021.12.16 10:48:14 INFO  time: indexed workspace in 6.86s
2021.12.16 10:48:15 INFO  compiling root (48 scala sources)
2021.12.16 10:48:16 INFO  no build target: using presentation compiler with only scala-library: 2.13.6
2021.12.16 10:48:27 INFO  compiling root-test (18 scala sources)
2021.12.16 10:48:27 INFO  time: compiled root in 11s
2021.12.16 10:48:29 INFO  time: compiled root-test in 2.15s

**Comment by gombosg** [comment]
I tried my project with downgrading to Metals 0.10.8 but no improvement happened.
Looks like @yaoqiang-se has the same 
```
 2021.12.16 10:44:16 INFO no build target: using presentation compiler with only scala-library: 3.1.0
```
issue.
We could maybe zoom in by learning what "build target" means and how it is determined?

@tgodzik any more debugging we could do?

**Comment by gombosg** [comment]
Going back to metals 0.10.5 worked! 🎉 Let me bisect the version where this breaks!

**Comment by gombosg** [comment]
@tgodzik confirmed! 0.10.5 works fine, with 0.10.6 or anything over, including snapshots, it doesn't work.

https://github.com/scalameta/metals/compare/v0.10.5..v0.10.6 🤔 

If you think some commits may be the culprit, tell me and I can bisect!

**Comment by gombosg** [comment]
@tgodzik 
This is the culprit after `git bisect`: https://github.com/scalameta/metals/commit/2422494d40951cb367e5c180e5c3621febfcd91f
```
~/p/o/metals ((0576d625…)|BISECTING|✔) $ git bisect good                                                                                                                                    
2422494d40951cb367e5c180e5c3621febfcd91f is the first bad commit
commit 2422494d40951cb367e5c180e5c3621febfcd91f
Author: Pavol Vidlicka <pvidlicka@samba.ai>
Date:   Thu Aug 5 18:52:55 2021 +0200

    Refactor FileWatcher to work on large project on macOS

 .../internal/metals/MetalsLanguageServer.scala     |  20 +-
 .../meta/internal/metals/watcher/FileWatcher.scala | 267 +++++++++++++--------
 .../meta/internal/metals/watcher/PathTrie.scala    |  73 ++++++
 3 files changed, 254 insertions(+), 106 deletions(-)
 create mode 100644 metals/src/main/scala/scala/meta/internal/metals/watcher/PathTrie.scala
~/p/o/metals ((0576d625…)|BISECTING|✔) $   
```

**Comment by tgodzik** [comment]
Interesting, thanks for digging into this! It would seem that file watcher is not notifying Metals about semanticdb files used for compilation. I wonder why is this problem popping up for you, but it's happening to other users :thinking:
